kernel_test.go

```diff
@@ -15,6 +16,249 @@ const (
+		}
+
+		msgParsed, _, err := WireMsgToComposedMsg(repMsgParts, []byte(connectionKey))
+		if err != nil {
+			t.Fatal("Could not parse wire message:", err)
+		}
+
+		ch <- msgParsed
+	}()
+
+	select {
+	case reply = <-ch:
+	case <-time.After(timeout):
+		t.Fatal("recvShellReply timed out")
+	}
+
+	return
+}
+
+// recvIOSub tries to read a published message from the IOPub channel. It will timeout after the given
+// timeout delay. Upon error or timeout, recvIOSub will Fail the test.
+func (client *testJupyterClient) recvIOSub(t *testing.T, timeout time.Duration) (sub ComposedMsg) {
+	t.Helper()
+
+	ch := make(chan ComposedMsg)
+
+	go func() {
+		repMsgParts, err := client.ioSocket.RecvMessageBytes(0)
+		if err != nil {
+			t.Fatal("IOPub socket RecvMessageBytes:", err)
+		}
+
+		msgParsed, _, err := WireMsgToComposedMsg(repMsgParts, []byte(connectionKey))
+		if err != nil {
+			t.Fatal("Could not parse wire message:", err)
+		}
+
+		ch <- msgParsed
+	}()
+
+	select {
+	case sub = <-ch:
+	case <-time.After(timeout):
+		t.Fatal("recvIOSub timed out")
+	}
+
+	return
+}
+
+// request preforms a request and awaits a reply on the shell channel. Additionally all messages on the IOPub channel
+// between the opening 'busy' messages and closing 'idle' message are captured and returned. The request will timeout
+// after the given timeout delay. Upon error or timeout, request will Fail the test.
+func (client *testJupyterClient) request(t *testing.T, request ComposedMsg, timeout time.Duration) (reply ComposedMsg, pub []ComposedMsg) {
+	t.Helper()
+
+	client.sendShellRequest(t, request)
+	reply = client.recvShellReply(t, timeout)
+
+	// Read the expected 'busy' message and ensure it is in fact, a 'busy' message
+	subMsg := client.recvIOSub(t, 1*time.Second)
+	assertMsgTypeEquals(t, subMsg, "status")
+
+	subData := getMsgContentAsJsonObject(t, subMsg)
+	execState := getString(t, "content", subData, "execution_state")
+
+	if execState != kernelBusy {
+		t.Fatalf("Expected a 'busy' status message but got '%v'", execState)
+	}
+
+	// Read messages from the IOPub channel until an 'idle' message is received
+	for {
+		subMsg = client.recvIOSub(t, 100*time.Millisecond)
+
+		// If the message is a 'status' message, ensure it is an 'idle' status
+		if subMsg.Header.MsgType == "status" {
+			subData = getMsgContentAsJsonObject(t, subMsg)
+			execState = getString(t, "content", subData, "execution_state")
+
+			if execState != kernelIdle {
+				t.Fatalf("Expected a 'idle' status message but got '%v'", execState)
+			}
+
+			// Break from the loop as we don't expect any other IOPub messages after the 'idle'
+			break
+		}
+
+		// Add the message to the pub collection
+		pub = append(pub, subMsg)
+	}
+
+	return
+}
+
+// assertMsgTypeEquals is a test helper that fails the test if the message header's MsgType is not the
+// expectedType.
+func assertMsgTypeEquals(t *testing.T, msg ComposedMsg, expectedType string) {
+	t.Helper()
+
+	if msg.Header.MsgType != expectedType {
```

Comment:
I wonder if this should just be inline to make it more readable?

Reply:
I was thinking the same but as it is already used 3 times and is likely to be used in future tests I figured it was nicer to avoid copy/pasting the message format everywhere. Then to clean up the format like we need to do with the `failure` char it's just one place that needs changing.